autoload/ale/lsp_linter.vim

```diff
@@ -407,9 +410,50 @@ endfunction
 " Clear LSP linter data for the linting engine.
 function! ale#lsp_linter#ClearLSPData() abort
     let s:lsp_linter_map = {}
+    let s:custom_callbacks_map = {}
 endfunction
 
 " Just for tests.
 function! ale#lsp_linter#SetLSPLinterMap(replacement_map) abort
     let s:lsp_linter_map = a:replacement_map
 endfunction
+
+function! s:HandleLSPResponseToCustomRequests(conn_id, response) abort
+    if has_key(a:response, 'id')
+    \&& has_key(s:custom_callbacks_map, a:response.id)
+        let l:Callback = remove(s:custom_callbacks_map, a:response.id)
+        call l:Callback(a:response)
+    endif
+endfunction
+
+function! s:OnReadyForCustomRequests(message, Callback, linter, lsp_details) abort
+    let l:id = a:lsp_details.connection_id
+    let l:Callback = function('s:HandleLSPResponseToCustomRequests')
+
+    call ale#lsp#RegisterCallback(l:id, l:Callback)
+
+    let l:request_id = ale#lsp#Send(l:id, a:message)
+    let s:custom_callbacks_map[l:request_id] = a:Callback
+endfunction
+
+" Send a custom request to an LSP linter.
+function! ale#lsp_linter#SendRequest(buffer, linter_name, method, parameters, Callback) abort
+    let l:filetype = ale#linter#ResolveFiletype(getbufvar(a:buffer, '&filetype'))
+    let l:linter_list = ale#linter#GetAll(l:filetype)
+    let l:linter_list = filter(l:linter_list, {_, v -> v.name is# a:linter_name})
+
+    if len(l:linter_list) < 1
+        throw 'Linter "' . a:linter_name . '" not found!'
+    endif
+
+    let l:linter = l:linter_list[0]
+
+    if empty(l:linter.lsp)
+        throw 'Linter "' . a:linter_name . '" does not support LSP!'
+    endif
+
+    let l:message = [0, a:method, a:parameters]
```

Comment:
I just realised that we'll need one more parameter for the value that is `0` here. That value is `1` when sending notifications, and JSON-RPC spec requires that notifications are sent without the `id` parameter used for other messages.

Maybe we could replace `method, parameters`, above with `message`, and document the format of the messages as `[is_notification, method, parameters]` in the help text. That will match the format returned by functions in `autoload/ale/lsp/message.vim`, some of which we might decide to expose later.

Reply:
We should make the callback optional for notifications, but required for messages that aren't notifications, as it shouldn't be stored in a Dictionary and won't be used. No responses are sent for notifications.

Reply:
Good point, I did not think of this use case. That also requires to make the handler argument optional, since no response is expected when a notification is sent.